File: library/src/main/java/de/timfreiheit/mozart/utils/QueueHelper.java

/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.timfreiheit.mozart.utils;

import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.List;

import de.timfreiheit.mozart.model.Playlist;

/**
 * Utility class to help on queue related tasks.
 */
public class QueueHelper {

    public static List<MediaSessionCompat.QueueItem> mediaQueueFromPlaylist(Playlist playlist) {

        List<MediaSessionCompat.QueueItem> queue = new ArrayList<>();
        List<MediaMetadataCompat> playlist1 = playlist.getPlaylist();
        for (int i = 0; i < playlist1.size(); i++) {
            MediaMetadataCompat track = playlist1.get(i);

            // We don't expect queues to change after created, so we use the item index as the
            // queueId. Any other number unique in the queue would work.
            MediaSessionCompat.QueueItem item = new MediaSessionCompat.QueueItem(
                    track.getDescription(), i);
            queue.add(item);

        }

        return queue;
    }

    public static int getMusicIndexOnQueue(Iterable<MediaSessionCompat.QueueItem> queue,
                                           String mediaId) {
        int index = 0;
        for (MediaSessionCompat.QueueItem item : queue) {
            if (mediaId.equals(item.getDescription().getMediaId())) {
                return index;
            }
            index++;
        }
        return -1;
    }

    public static int getMusicIndexOnQueue(Iterable<MediaSessionCompat.QueueItem> queue,
                                           long queueId) {
        int index = 0;
        for (MediaSessionCompat.QueueItem item : queue) {
            if (queueId == item.getQueueId()) {
                return index;
            }
            index++;
        }
        return -1;
    }

    public static boolean isIndexPlayable(int index, List<MediaSessionCompat.QueueItem> queue) {
        return (queue != null && index >= 0 && index < queue.size());
    }

    /**
     * Determine if two queues contain identical media id's in order.
     *
     * @param list1 containing {@link MediaSessionCompat.QueueItem}'s
     * @param list2 containing {@link MediaSessionCompat.QueueItem}'s
     * @return boolean indicating whether the queue's match
     */
    public static boolean equals(List<MediaSessionCompat.QueueItem> list1,
                                 List<MediaSessionCompat.QueueItem> list2) {
        if (list1 == list2) {
            return true;
        }
        if (list1 == null || list2 == null) {
            return false;
        }
        if (list1.size() != list2.size()) {
            return false;
        }
        for (int i = 0; i < list1.size(); i++) {
            if (list1.get(i).getQueueId() != list2.get(i).getQueueId()) {
                return false;
            }
            if (!TextUtils.equals(list1.get(i).getDescription().getMediaId(),
                    list2.get(i).getDescription().getMediaId())) {
                return false;
            }
        }
        return true;
    }

}

File: mozart/src/main/java/de/timfreiheit/mozart/utils/QueueHelper.kt
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.timfreiheit.mozart.utils

import android.support.v4.media.session.MediaSessionCompat
import android.text.TextUtils
import de.timfreiheit.mozart.model.Playlist
import java.util.*

/**
 * Utility class to help on queue related tasks.
 */

internal fun Playlist.createMediaQueue(): List<MediaSessionCompat.QueueItem> {

    val queue = ArrayList<MediaSessionCompat.QueueItem>()
    val playlist1 = playlist
    for (i in playlist1.indices) {
        val track = playlist1[i]

        // We don't expect queues to change after created, so we use the item index as the
        // queueId. Any other number unique in the queue would work.
        val item = MediaSessionCompat.QueueItem(
                track.description, i.toLong())
        queue.add(item)

    }

    return queue
}

internal fun List<MediaSessionCompat.QueueItem>.getMusicIndex(mediaId: String): Int {
    return this.indexOfFirst { mediaId == it.description.mediaId }
}

internal fun List<MediaSessionCompat.QueueItem>.getMusicIndex(queueId: Long): Int {
    return this.indexOfFirst { queueId == it.queueId }
}

internal fun List<MediaSessionCompat.QueueItem>?.isIndexPlayable(index: Int): Boolean {
    return this != null && index >= 0 && index < this.size
}

/**
 * Determine if two queues contain identical media id's in order.
 *
 * @param list1 containing [MediaSessionCompat.QueueItem]'s
 * @param list2 containing [MediaSessionCompat.QueueItem]'s
 * @return boolean indicating whether the queue's match
 */
internal fun areQueuesEqual(list1: List<MediaSessionCompat.QueueItem>?,
           list2: List<MediaSessionCompat.QueueItem>?): Boolean {
    if (list1 === list2) {
        return true
    }
    if (list1 == null || list2 == null) {
        return false
    }
    if (list1.size != list2.size) {
        return false
    }
    for (i in list1.indices) {
        if (list1[i].queueId != list2[i].queueId) {
            return false
        }
        if (!TextUtils.equals(list1[i].description.mediaId,
                list2[i].description.mediaId)) {
            return false
        }
    }
    return true
}

